frontend/main.js

```diff
@@ -103,19 +105,105 @@ async function startApp () {
+    const rootState = sbp('state/vuex/state')
+    if (rootState.loggedIn?.identityContractID !== identityContractID) return
+    if (!rootState.chatroom.currentChatRoomIDs[groupContractID] || rootState.chatroom.pendingChatRoomIDs[groupContractID] === chatRoomID) {
+      sbp('state/vuex/commit', 'setCurrentChatRoomId', { groupID: groupContractID, chatRoomID })
+    }
+  })
+
+  sbp('okTurtles.events/on', JOINED_GROUP, ({ identityContractID, groupContractID }) => {
+    const rootState = sbp('state/vuex/state')
+    if (rootState.loggedIn?.identityContractID !== identityContractID) return
+    if (!rootState[groupContractID]) return
+    if (!rootState.currentGroupId) {
+      sbp('state/vuex/commit', 'setCurrentGroupId', { contractID: groupContractID })
+      sbp('state/vuex/commit', 'setCurrentChatRoomId', {})
+    }
+  })
+
+  const switchCurrentChatRoomHandler = ({ identityContractID, groupContractID, chatRoomID }) => {
+    const rootState = sbp('state/vuex/state')
+    if (identityContractID && rootState.loggedIn?.identityContractID !== identityContractID) return
+    if (!rootState[groupContractID]) return
+    if (rootState.chatroom.currentChatRoomIDs[groupContractID] === chatRoomID) {
+      const id = rootState[groupContractID].generalChatRoomId || Object.entries(rootState[groupContractID].chatRooms).find(([id, value]) => {
+        // $FlowFixMe[incompatible-use]
+        return id !== chatRoomID && value.members[identityContractID]?.status === 'active'
+      })?.[0]
+      if (id) {
+        sbp('state/vuex/commit', 'setCurrentChatRoomId', { groupID: groupContractID, chatRoomID: id })
+      }
+    }
+  }
+
+  sbp('okTurtles.events/on', LEFT_CHATROOM, switchCurrentChatRoomHandler)
+  sbp('okTurtles.events/on', DELETED_CHATROOM, switchCurrentChatRoomHandler)
+
+  sbp('okTurtles.events/on', LEFT_GROUP, ({ identityContractID, groupContractID }) => {
+    const rootState = sbp('state/vuex/state')
+    console.error('@@@@@yy LEFT GROUP', identityContractID, groupContractID)
+    if (rootState.loggedIn?.identityContractID !== identityContractID) return
+    const state = rootState[identityContractID]
+    // grab the groupID of any group that we're a part of
+    const currentGroupId = rootState.currentGroupId
+    if (!currentGroupId || currentGroupId === groupContractID) {
+      const groupIdToSwitch = Object.keys(state.groups)
+        .filter(cID =>
+          cID !== groupContractID
+        ).sort(cID =>
+        // prefer successfully joined groups
+          sbp('state/vuex/state')[cID]?.profiles?.[groupContractID] ? -1 : 1
+        )[0] || null
+      sbp('state/vuex/commit', 'setCurrentChatRoomId', {})
+      sbp('state/vuex/commit', 'setCurrentGroupId', { contractID: groupIdToSwitch })
+      if (currentGroupId === groupContractID) {
+        sbp('controller/router').push({ path: '/' }).catch(() => {})
+      }
+    }
+  })
```

Comment:
`main.js` is not the right place for this sort of code. This belongs either in `controller/actions/chatroom.js` or in `model/contracts/chatroom.js`

Same for the rest of the code.

Reply:
Moved.